File: sais.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <algorithm>
#include <string>

using namespace std;

map<int, pair<int, int>> getBuckets(const vector<int>& T) {
    map<int, int> count;
    map<int, pair<int, int>> buckets;
    for (int c : T) count[c]++;
    int start = 0;
    for (auto& kv : count) {
        buckets[kv.first] = {start, start + kv.second};
        start += kv.second;
    }
    return buckets;
}

vector<int> sais(const vector<int>& T) {
    int n = T.size();
    vector<char> t(n, '_');
    t[n - 1] = 'S';
    for (int i = n - 1; i > 0; --i) {
        if (T[i - 1] == T[i]) t[i - 1] = t[i];
        else t[i - 1] = (T[i - 1] < T[i]) ? 'S' : 'L';
    }

    auto buckets = getBuckets(T);
    map<int, int> count;
    vector<int> SA(n, -1);
    map<int, int> LMS;
    int end = -1;
    for (int i = n - 1; i > 0; --i) {
        if (t[i] == 'S' && t[i - 1] == 'L') {
            int revoffset = ++count[T[i]];
            SA[buckets[T[i]].second - revoffset] = i;
            if (end != -1) LMS[i] = end;
            end = i;
        }
    }
    LMS[n - 1] = n - 1;

    count.clear();
    for (int i = 0; i < n; ++i) {
        if (SA[i] >= 0 && SA[i] > 0 && t[SA[i] - 1] == 'L') {
            int symbol = T[SA[i] - 1];
            int offset = count[symbol]++;
            SA[buckets[symbol].first + offset] = SA[i] - 1;
        }
    }

    count.clear();
    for (int i = n - 1; i > 0; --i) {
        if (SA[i] > 0 && t[SA[i] - 1] == 'S') {
            int symbol = T[SA[i] - 1];
            int revoffset = ++count[symbol];
            SA[buckets[symbol].second - revoffset] = SA[i] - 1;
        }
    }

    vector<int> namesp(n, -1);
    int name = 0;
    int prev = -1;
    for (int i = 0; i < n; ++i) {
        if (SA[i] >= 0 && SA[i] > 0 && t[SA[i]] == 'S' && t[SA[i] - 1] == 'L') {
            if (prev != -1) {
                int a_start = SA[prev], a_end = LMS.count(SA[prev]) ? LMS[SA[prev]] : n;
                int b_start = SA[i], b_end = LMS.count(SA[i]) ? LMS[SA[i]] : n;
                vector<int> a(T.begin() + a_start, T.begin() + a_end);
                vector<int> b(T.begin() + b_start, T.begin() + b_end);
                if (a != b) name++;
            }
            prev = i;
            namesp[SA[i]] = name;
        }
    }

    vector<int> names, SApIdx;
    for (int i = 0; i < n; ++i) {
        if (namesp[i] != -1) {
            names.push_back(namesp[i]);
            SApIdx.push_back(i);
        }
    }

    if (name < (int)names.size() - 1) {
        names = sais(names);
    }

    reverse(names.begin(), names.end());

    SA.assign(n, -1);
    count.clear();
    for (int i = 0; i < (int)names.size(); ++i) {
        int pos = SApIdx[names[i]];
        int revoffset = ++count[T[pos]];
        SA[buckets[T[pos]].second - revoffset] = pos;
    }

    count.clear();
    for (int i = 0; i < n; ++i) {
        if (SA[i] >= 0 && SA[i] > 0 && t[SA[i] - 1] == 'L') {
            int symbol = T[SA[i] - 1];
            int offset = count[symbol]++;
            SA[buckets[symbol].first + offset] = SA[i] - 1;
        }
    }

    count.clear();
    for (int i = n - 1; i > 0; --i) {
        if (SA[i] > 0 && t[SA[i] - 1] == 'S') {
            int symbol = T[SA[i] - 1];
            int revoffset = ++count[symbol];
            SA[buckets[symbol].second - revoffset] = SA[i] - 1;
        }
    }

    return SA;
}

vector<int> string_to_intvec(const string& s) {
    vector<int> T;
    for (char c : s) T.push_back((int)c);
    return T;
}

int main(int argc, char* argv[]) {
    string text;
    if (argc < 2) {
        text = "GTCCCGATGTCATGTCAGGA$";
    } else {
        ifstream file(argv[1], ios::in | ios::binary);
        if (!file) {
            cerr << "Error: Cannot open file " << argv[1] << endl;
            return 1;
        }
        text.assign((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
        text += "$";
    }
    vector<int> T = string_to_intvec(text);
    vector<int> SA = sais(T);
    cout << "Suffix array size: " << SA.size() << endl;
    // Uncomment to print the suffix array
    // for (int i : SA) cout << i << " ";
    // cout << endl;
    return 0;
}

File: sais.py
import time
import sys

def getBuckets(T):
    count = {}
    buckets = {}
    for c in T:
        count[c] = count.get(c, 0) + 1
    start = 0
    for c in sorted(count.keys()):
        buckets[c] = [start, start + count[c]]  # Solo cambio: usar lista en vez de tupla
        start += count[c]
    return buckets

def sais(T):
    if len(T) <= 1:
        return list(range(len(T)))
        
    t = ["_"] * len(T)
    t[len(T) - 1] = "S"
    
    for i in range(len(T) - 2, -1, -1):
        if T[i] == T[i + 1]:
            t[i] = t[i + 1]
        else:
            t[i] = "S" if T[i] < T[i + 1] else "L"
    
    buckets = getBuckets(T)
    count = {}
    SA = [-1] * len(T)
    LMS = {}
    end = None
    
    for i in range(1, len(T)):
        if t[i] == "S" and t[i - 1] == "L":
            revoffset = count[T[i]] = count.get(T[i], 0) + 1
            SA[buckets[T[i]][1] - revoffset] = i
            if end is not None:
                LMS[i] = end
            end = i

    LMS[len(T) - 1] = len(T) - 1
    count.clear()  # Cambio: usar clear() en vez de reasignar
    
    for i in range(len(T)):
        if SA[i] > 0:
            if t[SA[i] - 1] == "L":
                symbol = T[SA[i] - 1]
                offset = count.get(symbol, 0)
                SA[buckets[symbol][0] + offset] = SA[i] - 1
                count[symbol] = offset + 1

    count.clear()  # Cambio: usar clear() en vez de reasignar
    for i in range(len(T) - 1, -1, -1):
        if SA[i] > 0:
            if t[SA[i] - 1] == "S":
                symbol = T[SA[i] - 1]
                revoffset = count[symbol] = count.get(symbol, 0) + 1
                SA[buckets[symbol][1] - revoffset] = SA[i] - 1

    return SA

def get_bwt(text, sa):
    """Construye la BWT usando el arreglo de sufijos"""
    bwt = []
    for i in sa:
        if i == 0:
            bwt.append(text[-1])
        else:
            bwt.append(text[i - 1])
    return ''.join(bwt)

def get_first_column(bwt):
    """Obtiene la primera columna ordenando la BWT"""
    return ''.join(sorted(bwt))

def get_counts(bwt):
    """Calcula la tabla C (cuenta de caracteres menores)"""
    counts = {}
    total = 0
    # Primero contamos frecuencias
    for c in sorted(set(bwt)):
        counts[c] = bwt.count(c)
    # Luego calculamos las posiciones acumuladas
    running_sum = 0
    final_counts = {}
    for c in sorted(counts.keys()):
        final_counts[c] = running_sum
        running_sum += counts[c]
    return final_counts

def get_occ(bwt, char, pos):
    """Cuenta ocurrencias del carácter hasta la posición dada"""
    return bwt[:pos].count(char)

def fm_search(pattern, text, sa):
    """Busca un patrón usando FM-Index"""
    bwt = get_bwt(text, sa)
    first_col = get_first_column(bwt)
    counts = get_counts(bwt)
    
    # Inicializar rango de búsqueda
    start = 0
    end = len(bwt)
    
    # Buscar el patrón de derecha a izquierda
    for c in reversed(pattern):
        if c not in counts:
            return []
        
        # Actualizar rango usando LF-mapping
        start = counts[c] + get_occ(bwt, c, start)
        end = counts[c] + get_occ(bwt, c, end)
        
        if start >= end:
            return []
    
    # Convertir posiciones BWT a posiciones en el texto original usando SA
    return [sa[i] for i in range(start, end)]

def main():
    if len(sys.argv) != 2:
        print("Uso: python sais.py <archivo_texto>")
        sys.exit(1)
        
    try:
        with open(sys.argv[1], 'r', encoding='utf-8') as file:
            text = file.read().strip() + '$'
            
        print(f"Procesando archivo: {sys.argv[1]}")
        print(f"Tamaño del texto: {len(text)} caracteres")
        
        T = [ord(c) for c in text]
        
        start_time = time.time()
        SA = sais(T)
        end_time = time.time()
        
        print(f"Tiempo de ejecución: {end_time - start_time:.2f} segundos")
        
        # Ejemplo de búsqueda
        pattern = input("Ingrese el patrón a buscar: ")
        positions = fm_search(pattern, text, SA)
        
        if positions:
            print(f"Patrón '{pattern}' encontrado en las posiciones: {positions}")
        else:
            print(f"Patrón '{pattern}' no encontrado en el texto")
        
        with open(f"{sys.argv[1]}_SA.txt", 'w') as f:
            f.write(','.join(map(str, SA)))
            
    except FileNotFoundError:
        print(f"Error: No se encuentra el archivo {sys.argv[1]}")
    except Exception as e:
        print(f"Error: {str(e)}")

if __name__ == "__main__":
    main()